Patch so master to master

joefitter opened the pull request.

joefitter committed:
use so branch of hmpo-form-wizard

joefitter committed:
Enhancement - pass rawValue to template in confirm controller (#80)

* Enhancement - pass rawValue to template in confirm controller

Modifier functions can be passed to confirm controller to format or translate values, but it would be useful to be able to access the raw unformatted value.

* pass new rawValue variable to template which is not put through modifier function
* updated README
* added test

* rawValue > origValue. Also refactored to call modifier with scope of field object

joefitter committed:
bugfix - don't include empty values in confirm controller

For some reason we were filtering out empy values, except for when there is a modifier for that value. I can't see a use case where we would desire this behaviour, and it results in the modifier being called on the empy field and included in the results.

easternbloc committed:
Merge pull request #82 from UKHomeOffice/bugfix/dont-include-empty-values

bugfix - don't include empty values in confirm controller

joefitter committed:
Enhancement - generate confirm table data from steps

* added logic to generate an array of sections containing fields for display on confirm page and email
* updated readme
* added tests

joefitter committed:
typo

joefitter committed:
Feature - integrate hof emailer

* added hof-emailer
* refactored confirm controller to pass formatted data to emailer
* added tests

joefitter committed:
proxyquire issue

easternbloc committed:
Merge pull request #86 from UKHomeOffice/enhancement/integrate-hof-emailer

Enhancement/integrate hof emailer

joefitter committed:
merged in master

joefitter committed:
fix version of hmpo-form-wizard

joefitter commented:
@JoeChapman @easternbloc @danielepolencic @bendiggle @gxxm mind having a 👀 - patching so-master back to master. Contains the changes to the confirm controller and integration with hof-emailer

JoeChapman commented on package.json:
```diff
@@ -17,6 +17,7 @@
   },
   "dependencies": {
     "hmpo-form-wizard": "5.1.1",
+    "hof-emailer": "^1.0.0",
```
there is a newer version of hof-emailer (1.1.0). Can this be updated?

joefitter replied:
this is just bringing master up to date with so-master, so no changes have actually been made to this branch. I would suggest bumping hof-emailer in a separate PR

JoeChapman replied:
ok

JoeChapman commented on test/lib/confirm-controller.js:
```diff
@@ -1,126 +1,409 @@
+        './util/helpers': helpersStub,
+        'hof-emailer': EmailerStub
+      });
+      const steps = {
```
isn't this another fixture? Most other fixtures in this PR have their own files

joefitter replied:
yeah I guess we could split it out, where do we draw the line though - do we add the field config to a separate file  (7 lines). The idea of fixtures were more real use case like scenarios, whereas these 'fixtures' are testing a simple use case

danielepolencic commented on lib/util/helpers.js:
```diff
@@ -0,0 +1,107 @@
+        if (result === path) {
+          return key;
+        }
+      }
```
This is quite hard to follow. Is there any way we could rather write something like:

``` javascript
let result = [
  `${base}.${key}.summary`,
  `${base}.${key}.label`,
  `${base}.${key}.legend`,
  `${base}.${key}.header`
].some(path => {
  return translate(path) === path;
});
return result ? key : undefined;
```

joefitter replied:
that's much nicer

danielepolencic replied:
I haven't tested the code though.

joefitter replied:
@danielepolencic I haven't updated this yet as up to my eyeballs in project work, but there is a potentially nicer solution used elsewhere in the architecture - i18n-lookup (one of Lenny's packages) - pass in an array of keys and it returns the first one found

danielepolencic commented on test/lib/confirm-controller.js:
```diff
@@ -1,126 +1,409 @@
-          fields: ['field-one', 'field-two']
+    StubController.prototype.locals = sinon.stub().returns({});
+    StubController.prototype.get = sinon.stub();
+    EmailerStub.prototype.sendEmails = sinon.stub().returns(new Promise(resolve => resolve()));
```
`new Promise(resolve => resolve())` is `Promise.resolve()`

joefitter replied:
magic!

joefitter committed:
stash

joefitter committed:
Feature - implied title and intro (#91)

This change extends the locals method of baseController providing route (without preceeding slash), title and intro.

Title is looked up in the following order:
```js
pages.{routeName}.header -> fields.{firstFieldName}.label -> fields.{firstFieldName}.legend
```

Intro is looked up using `pages.{routeName}.intro`

Added i18n-lookup lib for translation lookups, added mustache for interpolation in translations

joefitter committed:
bugfix - ignore duplicate fields

If a field appears on multiple steps in the same section, only the first one should be used. This is a bug which occurs when a journey has a potential alternate step containaing one or more fields that could be in a different part of the journey

joefitter committed:
typo

joefitter committed:
added test to check for dulplicates

JoeChapman committed:
bugfix - ignore duplicate fields (#93)

* bugfix - ignore duplicate fields

If a field appears on multiple steps in the same section, only the first one should be used. This is a bug which occurs when a journey has a potential alternate step containaing one or more fields that could be in a different part of the journey

* typo

* added test to check for dulplicates

joefitter committed:
bugfix - format data on post

* due to multiple instances data needs to be formatted on get (display) and post (email)

JoeChapman committed:
Merge pull request #94 from UKHomeOfficeForms/bugfix/format-data-on-post

bugfix - format data on post

joefitter committed:
bugfix - fix getStepNameFromField helper

* pass visited steps list
* only pick step from list of visited steps

joefitter committed:
Merge pull request #95 from UKHomeOfficeForms/bugfix/fix-for-field-on-multipl-steps

bugfix - fix getStepNameFromField helper

joefitter committed:
enhancement - allow email subject to be an object

req.translate uses deepTranslate middleware which attempts to resolve objects returned from translations. Raw translate will return the object, this is useful if we have a different subject for caseworker and customer

JoeChapman committed:
Merge pull request #96 from UKHomeOfficeForms/enhancement/allow-subject-to-be-object

enhancement - allow email subject to be an object

joefitter commented:
@JoeChapman @easternbloc @danielepolencic @bendiggle mind having a look over this again please? @danielepolencic - changes made as suggested. This now also contains some more changes mainly to the confirm controller

joefitter commented:
ignore - rebase fail. Will resubmit!
